### ISSUE TITLE
Support xcopy-esque venv installation

### ISSUE THREAD

**Issue by aaronsteers** [meltano/meltano#7004]
## Assumptions

1. We're running on Docker with explicit control of the base image.
1. Meltano is already installed.
1. The container may or may not have a `.meltano` folder created as of yet.
1. We have cached one or more venv folder(s) running on the same exact Docker image (meaning: also same exact version of Pip+Python).
1. We are controlling for changes in `pip_url` via a hash or other constraint.

## Requirements

1. Assuming all of the above have been controlled for, allow 'near-instant' plugin installation via xcopy into the correct location. **OR:**
2. Allow Meltano to read from a different already-existing directory when searching for a specific plugin's `venv`.

## Related

- The GitHub `python-setup` action proves this is possible, with near-instant bootup of projects and no installation needed when the precheck of the installation definition hashes matches an existing cache.
- This tackles a similar challenge as #2089 and related issues, but instead of skipping installs we try to make them near-instantaneous.

**Comment by aaronsteers** [comment]
@z3z1ma - I think you mentioned you have a working implementation via PEX? Is that right?

https://pex.readthedocs.io/en/v2.1.123/

**Comment by z3z1ma** [comment]
@aaronsteers yes I do. I actually plan to drop the project on github soon so you can dig in yourself!

**Comment by aaronsteers** [comment]
Very timely: Dagster is also moving to PEX for some of their Python installs.

https://dagster.io/blog/fast-deploys-with-pex-and-docker

>  What is PEX?
> Short for Python Executable, [pex](https://pex.readthedocs.io/) is a tool that bundles Python packages into files called pex files. These are executable files that carry Python packages and some bootstrap code within them...
> 
> Using pex in combination with S3 for storing the pex files, we built a system where the fast path avoids the overhead of building and launching Docker images.

**Comment by aaronsteers** [comment]
I got this working today with `shiv`, another alternative to `pex.

Prereqs: a valid Meltano project, with `meltano` and `pipx` installed

### Steps to reproduce

## 0 - Setup some vars

Just so the later code can be generic and easier to read:

```
export PLUGIN_NAME=target-duckdb
export PIP_URL=target-duckdb
```

### Step 1: Build the `pyz` file.

```sh
pipx run shiv -c $PLUGIN_NAME -o ./$PLUGIN_NAME.pyz $PIP_URL
```

Which resolves to:

```console
pipx run shiv -c target-duckdb -o ./target-duckdb.pyz target-duckdb
```

_Note: Using `pipx run`, we don't actually need to preinstall `shiv`. Alternatively, you could pip install it and then invoke it directly._

### Step 2: Test the executable

```console
aj@ajs-macbook-pro jaffle-shop-template % ./target-duckdb --help    
usage: target-duckdb [-h] [-c CONFIG]

options:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        Config file
```

### Step 3: Update `meltano.yml`

In `meltano.yml`, add a new `executable` entry with the absolute path:

```yml
# ...
  loaders:
  - name: target-duckdb
    variant: jwills
    pip_url: target-duckdb~=0.4
    executable: /Users/aj/Source/jaffle-shop-template/target-duckdb
# ...
```

### Step 4: Invoke with Meltano

```
aj@ajs-macbook-pro jaffle-shop-template %  meltano invoke target-duckdb --help
2023-03-12T22:08:16.433582Z [info     ] Environment 'dev' is active
usage: target-duckdb [-h] [-c CONFIG]

options:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        Config file
```

This works even if you've not yet executed `meltano install`. 🚀 

## Notes / Takeaways

1. `.pyz` has a PEP associated with it, and is a recommended extension for python executables.
2. There was no way I could find to nullify the `pip_url` to make it skip the plugin during `meltano install`.
3. I only got this working using absolute paths.
4. The extension doesn't matter. You can also create the file with no extension, and it still works.
5. Files size was 15MB for `target-duckdb` (installed from PyPi) and 43MB for `target-jaffle-shop` (installed from source).
6. I noticed a ~2 second slowdown on first invocation (probably during initial decompression) and then execution times seemed identical.

**Comment by WillDaSilva** [comment]
Comparison of Shiv and pex: https://shiv.readthedocs.io/en/latest/history.html#motivation-comparisons

**Comment by visch** [comment]
https://discuss.python.org/t/pip-plans-to-introduce-an-alternative-zipapp-deployment-method/17431/2 writeup about PIP @aaronsteers

**Comment by pnadolny13** [comment]
I brought this up in office hours but it might be helpful to split our pip_url into an explicit list vs a space separated string. dbt implements their package in a similar way, see https://docs.getdbt.com/docs/build/packages:

```yaml
packages:
  - package: dbt-labs/snowplow
    version: 0.7.0

  - git: "https://github.com/dbt-labs/dbt-utils.git"
    revision: 0.9.2

  - local: /opt/dbt/redshift
 ```

This could help for the issue that AJ brought up where you have a mix of packages, some pex, some git urls, some etc. It also could help with setting version ranges in the future.
